Add an elementwise + activation fusion pass, add mechanizm to add attributes outside of base op.

tsocha opened the pull request:
### PR types
Performance optimization

### PR changes
Others

### Describe
Add a pass which fuse element wise operation with an activation:
- "elementwise_add", "elementwise_sub", "elementwise_mul"
- "relu", "tanh", "leaky_relu", "swish", "hardswish", "sqrt", "abs", "clip", "gelu", "relu6", "sigmoid"
Add a mechanizm to add attributes outside of base op:
- Op replacement pass
- New OneDNN specific operators which extend base ops by new attributes.

jczaja commented:
@piotrekobiIntel ,@Silv3S Please review

jakpiase commented on paddle/fluid/framework/ir/mkldnn/elt_act_onednn_fuse_pass.cc:
```diff
@@ -0,0 +1,139 @@
+            .LE("elementwise_sub", 1)
+            .LE("elementwise_mul", 1)
+            .LE("elementwise_div", 1)
+            .LE("elementwise_pow", 1)
```
Are you sure that we should check elementwise_pow and elementwise_div versions if we don't have kernels for these ops?

jakpiase commented on paddle/fluid/framework/ir/mkldnn/elt_act_onednn_fuse_pass.cc:
```diff
@@ -0,0 +1,140 @@
+using string::PrettyLogDetail;
+
+void ElementwiseActivationOneDNNPass::ApplyImpl(Graph *graph) const {
+  std::vector<std::string> act_types = {"relu",  "tanh",      "leaky_relu",
```
Sigmoid, relu6 and "scale(linear activation)" can be also a part of this fuse, what do you think about adding them?

tsocha replied:
I will try to do that :)

arlesniak replied:
Maybe enum would be better than strings especially during later comparison. If there are common maps/enums of the act_types, ew_types etc. they can be used in lines 134-145 as well in algo_map elementwise_mkldnn_op.h:41 . What is your opinion ?

tsocha replied:
I have a problem with enums because it is saved as a op attribute, such attribute have to be reported in the `compat`'s pbtxt files. I wonder how I can report it there if it is not a standard type.

jakpiase commented on paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h:
```diff
@@ -32,6 +32,34 @@ using mkldnn::stream;
+        {"sqrt", dnnl::algorithm::eltwise_sqrt},
+        {"abs", dnnl::algorithm::eltwise_abs},
+        {"clip", dnnl::algorithm::eltwise_clip},
+        {"gelu", dnnl::algorithm::eltwise_gelu_erf}};
```
That is actually not correct, sometimes we are using "gelu_tanh", that depends on the "approximate" attribute of gelu op

jakpiase commented on python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_elt_act_fuse_pass.py:
```diff
@@ -0,0 +1,392 @@
+        self.pass_name = 'elt_act_onednn_fuse_pass'
+
+
+class ElementwiseActivationMkldnnFusePassTest_Div_Relu(ElementwiseActivationMkldnnFusePassTest):
```
Do we need elementwise_div and elementwise_pow tests if we do not support these operators?

jakpiase commented on python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_elt_act_fuse_pass.py:
```diff
@@ -0,0 +1,392 @@
+        self.pass_name = 'elt_act_onednn_fuse_pass'
+
+
+class ElementwiseActivationMkldnnFusePassTest_Add_Gelu(ElementwiseActivationMkldnnFusePassTest):
```
Please, add also a test in which "gelu" will use approximate version of algorithm (dnnl::algorithm::gelu_tanh)

arlesniak commented on python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_elt_act_fuse_pass.py:
```diff
@@ -0,0 +1,392 @@
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Please add `paddle.enable_static()`

arlesniak commented on python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_elt_act_fuse_pass.py:
```diff
@@ -0,0 +1,392 @@
+    def set_params(self):
+        self.operand = fluid.layers.elementwise_add
+        self.act = fluid.layers.relu
+        self.pass_name = 'elt_act_onednn_fuse_pass'
```
You could move `self.pass_name = 'elt_act_onednn_fuse_pass'` to line 31 in `setUp`, so you wouldn't have to repeat that line in  every overriden `set_params` method. What is your opinion?

arlesniak commented on python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_elt_act_fuse_pass.py:
```diff
@@ -0,0 +1,392 @@
+        self.check_output_with_option(use_gpu)
+
+    def test_pass_compatible(self):
+        # import pudb
```
Please remove commented out code

arlesniak commented on paddle/fluid/framework/ir/mkldnn/elt_act_onednn_fuse_pass.cc:
```diff
@@ -0,0 +1,140 @@
+
+    auto *elementwise_op = elementwise->Op();
+
+    const std::string wo_elt_type =
```
It looks like `wo_elt_type` is used only inside `if (elementwise_op->HasAttr("use_mkldnn"))` code block so it should be moved in there. What is your opinion?

arlesniak commented:
Looks good, but I've left a comment

lidanqing-vv commented on paddle/fluid/operators/elementwise/elementwise_op.h:
```diff
@@ -185,6 +185,26 @@ class ElementwiseOpMaker : public framework::OpProtoAndCheckerMaker {
+    AddAttr<std::string>("activation_type",
+                         "Activation type used in elementwise operator.")
+        .SetDefault("")
+        .AsExtra();
```
You may add supported activation here like this?
`.InEnum({"float32", "int8", "bfloat16"})`

tsocha replied:
If it is an enum it have to be reported here as well: https://github.com/PaddlePaddle/Paddle/pull/36541/files#diff-f4d5cae745e1a49dddc2e9b55aeb92f4d98b3a3093e902ae742af1517c28bcb2

Do you know how to do it?

lidanqing-vv replied:
Hi, I don't mean it should be enum type, I mean it may be like     
```
AddAttr<std::string>(
        "mkldnn_data_type",
        "(string, default \"float32\"). Data type of mkldnn kernel")
        .SetDefault("float32")
        .InEnum({"float32", "int8", "bfloat16"})
        .AsExtra();
```
But now I saw fc_mkldnn_op.cc also
```
 AddAttr<std::string>("activation_type",
                         "Activation type used in elementwise operator.")
        .SetDefault("")
```
So it doesn't have to be added

jakpiase commented on paddle/fluid/operators/compat/elementwise_sub.pbtxt:
```diff
@@ -35,6 +35,10 @@ extra {
     type: FLOAT
   }
+  attrs {
+    name: "activation_scale"
```
What is that attr used for?

tsocha replied:
https://docs.oneapi.io/versions/latest/onednn/dev_guide_attributes_post_ops.html#doxid-dev-guide-attributes-post-ops-1dev-guide-attributes-post-ops-eltwise

it is used in the INT8 inference, I added it here in advance because it is used in onednn's post_ops.

jakpiase replied:
Ok, now it is clear, thank you!

lidanqing-vv commented on paddle/fluid/inference/api/paddle_pass_builder.cc:
```diff
@@ -253,7 +253,7 @@ void CpuPassStrategy::EnableMKLDNN() {
              // "fc_mkldnn_pass",
              // "fc_act_mkldnn_fuse_pass",
-             "batch_norm_act_fuse_pass",
+             "batch_norm_act_fuse_pass", "elt_act_onednn_fuse_pass",
```
Hi, Have you installed pre-commit which will automatically format the code. Maybe this will be added automatically in next line.

tsocha replied:
It was added in the next line by me but a code formater changed it this way.
I think this behavior is caused by these commented lines of code.

lidanqing-vv replied:
@tsocha  Hi, could you please commit it again under paddle official docker?  Please try this. Because all Baidu Paddle developers are now work under docker. Cause they are customers let's make sure to get aligned with Baidu. Baoachun mentioned this line many times already, only we made sure under Baidu official docker it is still like this, we can proceed more converstion with them.

lidanqing-vv commented:
Hi, @baoachun Could you please contact some of following people to approve this PR. 

This PR added elementwise+many activations fuse because now oneDNN support such big fuse. 

This PR changed op attributes and pbtxt file, and below logs seem showing some document md5 is changed. Thanks
```
2021-10-25 17:11:13 ****************
2021-10-25 17:11:13 Please find RD for approval first, and then find TPM for approval.
2021-10-25 17:11:13 0. You must have  one TPM approval for API documents change:
2021-10-25 17:11:13  jzhang533/ZhangJun, dingjiaweiww/DingJiaWei, Heeenrrry/LiKunLun, TCChenlong/ChenLong for general API docs
2021-10-25 17:11:13  PangHua/XiangHui for distributed related API docs
2021-10-25 17:11:13  twismon/WangYunKai, CheQiXiao/CheQiXiao for inference related API docs.
2021-10-25 17:11:13
2021-10-25 17:11:13 There are 1 approved errors.
2021-10-25 17:11:13 ****************
2021-10-25 17:11:14 API Difference is:
2021-10-25 17:11:14 - paddle.Tensor.add (ArgSpec(args=['x', 'y', 'name'], varargs=None, varkw=None, defaults=(None,), kwonlyargs=[], kwonlydefaults=None, annotations={}), ('document', 'bca0b1e5d03db0e546f54de21974f772'))
2021-10-25 17:11:14 ?                                                                                                                                                                      ^^^ ----  ^  ^^^ ^^^  ^ ^^^^^^
2021-10-25 17:11:14
2021-10-25 17:11:14 + paddle.Tensor.add (ArgSpec(args=['x', 'y', 'name'], varargs=None, varkw=None, defaults=(None,), kwonlyargs=[], kwonlydefaults=None, annotations={}), ('document', '6b4c8d613d2110fed8f2e282c08ddd9b'))
2021-10-25 17:11:14 ?                                          
```

baoachun commented:
Hi @tsocha @lidanqing-intel , I will find relevant people for approve later.

Do file names and pass names need to be unified?

tsocha commented:
@baoachun 
> 
> Hi @tsocha @lidanqing-intel , I will find relevant people for approve later.
> 
> Do file names and pass names need to be unified?

I think it is not necessary that these names match, it is only important that it is not misleading IMHO.
If it is not a problem I would to leave it as is.

jczaja commented:
@baoachun We have a difficulty passing PR-CI-Op-Benchmark. It does not seem to be a problem caused by this PR yet multiple reruns did not brought success. Please advice

baoachun commented:
Hi @jczaja , I see PR-CI-OP-benchmark is still queuing, but PR-CI-Static-Check failed, probably because the pre-commit unified code format is not used.

tsocha commented:
> 
> 
> Hi @jczaja , I see PR-CI-OP-benchmark is still queuing, but PR-CI-Static-Check failed, probably because the pre-commit unified code format is not used.

@baoachun please use this link for op_benchmark logs: https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/detail/3965092/job/8869286

baoachun commented:
Hi @jczaja , I see PR-CI-OP-benchmark is still queuing, but PR-CI-Static-Check failed, probably because the pre-commit unified code format is not used.

> > Hi @jczaja , I see PR-CI-OP-benchmark is still queuing, but PR-CI-Static-Check failed, probably because the pre-commit unified code format is not used.
> 
> @baoachun please use this link for op_benchmark logs: https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/detail/3965092/job/8869286

Got it, I will check it out tomorrow.

tsocha commented:
@baoachun did you have time to look at **PR-CI-OP-benchmark**?
Could you advise me how to fix it?

baoachun commented:
> @baoachun did you have time to look at **PR-CI-OP-benchmark**? Could you advise me how to fix it?

By looking at the failure log, I found that the reason for the failure was a decline in the performance of a single operator.
<img width="847" alt="图片" src="https://user-images.githubusercontent.com/22114318/140917941-fc796a34-c968-443f-a014-f66dc39af251.png">

tsocha commented:
> 
> 
> > @baoachun did you have time to look at **PR-CI-OP-benchmark**? Could you advise me how to fix it?
> 
> By looking at the failure log, I found that the reason for the failure was a decline in the performance of a single operator. <img alt="图片" width="847" src="https://user-images.githubusercontent.com/22114318/140917941-fc796a34-c968-443f-a014-f66dc39af251.png">

@baoachun I changed the **CPU** code but the test which failed is a benchmark of Add Backward on **GPU**.
I just added new attributes to the base operator.
Is it possible that **Add Backward GPU** performance is affected by that?

baoachun commented:
Hi @tsocha,   the reason for the failure may be that the latest code was not pulle. You can merge the latest code and submit again, if it still fails, I will ask a colleague for help.

lidanqing-vv commented:
@tsocha Could you please rebase/pull newest develop and solve the conflicts ? Then it may pass.

tsocha commented:
> 
> 
> Hi @tsocha, the reason for the failure may be that the latest code was not pulle. You can merge the latest code and submit again, if it still fails, I will ask a colleague for help.

@baoachun 
I'm merging _develop_ branch to my code daily, unfortunately **PR-CI-OP-benchmark** didn't pass.

I believe that this failure is not connected with my changes directly.
I think the reason why it fails is that GPU plugin iterate over attributes and since I added some of them the overall framework overhaul increased.
The failing test have a little data to compute so this extra overhaul caused a performance degradation in this case.
It should not impact the bigger production models IMHO.

jakpiase commented on paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.cc:
```diff
@@ -0,0 +1,148 @@
+    else
+      elementwise_op->SetAttr("activation_type", act_type);
+
+    elementwise_op->SetAttr("use_mkldnn", true);
```
eltwise op must have use_mkldnn attribute already set to `true`, it is enforced in lines 87-95, otherwise an exception will be thrown, so this line is redundant

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_add_op.cc:
```diff
@@ -127,6 +127,16 @@ REGISTER_OP_CPU_KERNEL(
+    elementwise_add_one_dnn,
+    ops::ElementwiseAddKernel<paddle::platform::CPUDeviceContext, float>,
+    ops::ElementwiseAddKernel<paddle::platform::CPUDeviceContext, double>,
+    ops::ElementwiseAddKernel<paddle::platform::CPUDeviceContext, int>,
```
Do we support `double`, `int` and `int64_t` datatypes?

tsocha replied:
It's the registration of the reference kernel, there can be models where the original operator was replaced with the custom one but since it's type is double it will fallback to reference.

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_mul_one_dnn_op.cc:
```diff
@@ -0,0 +1,85 @@
+  void AddInputX() override {
+    AddInput("X",
+             "(Variable), Tensor or LoDTensor of any dimensions. Its dtype "
+             "should be int32, int64, float32, float64.");
```
As above, do we support other dtypes than `float32`?

tsocha replied:
https://github.com/PaddlePaddle/Paddle/pull/36541#discussion_r797554318

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_mul_one_dnn_op.cc:
```diff
@@ -0,0 +1,85 @@
+  void AddInputY() override {
+    AddInput("Y",
+             "(Variable), Tensor or LoDTensor of any dimensions. Its dtype "
+             "should be int32, int64, float32, float64.");
```
As above

tsocha replied:
https://github.com/PaddlePaddle/Paddle/pull/36541#discussion_r797554318

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_mul_op.cc:
```diff
@@ -171,6 +171,17 @@ REGISTER_OP_CPU_KERNEL(
+    ops::ElementwiseMulKernel<paddle::platform::CPUDeviceContext, float>,
+    ops::ElementwiseMulKernel<paddle::platform::CPUDeviceContext, double>,
+    ops::ElementwiseMulKernel<paddle::platform::CPUDeviceContext, int>,
+    ops::ElementwiseMulKernel<paddle::platform::CPUDeviceContext, int64_t>,
```
As with other eltwises, do we support double, int and int64_t datatypes?

tsocha replied:
https://github.com/PaddlePaddle/Paddle/pull/36541#discussion_r797554318

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_sub_one_dnn_op.cc:
```diff
@@ -0,0 +1,86 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_sub_one_dnn_op.cc:
```diff
@@ -0,0 +1,86 @@
+  void AddInputX() override {
+    AddInput("X",
+             "(Variable), Tensor or LoDTensor of any dimensions. Its dtype "
+             "should be int32, int64, float32, float64.");
```
As with other eltwises, do we support double, int and int64_t datatypes?

tsocha replied:
https://github.com/PaddlePaddle/Paddle/pull/36541#discussion_r797554318

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_sub_one_dnn_op.cc:
```diff
@@ -0,0 +1,86 @@
+  void AddInputY() override {
+    AddInput("Y",
+             "(Variable), Tensor or LoDTensor of any dimensions. Its dtype "
+             "should be int32, int64, float32, float64.");
```
As above

tsocha replied:
https://github.com/PaddlePaddle/Paddle/pull/36541#discussion_r797554318

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_sub_one_dnn_op.h:
```diff
@@ -0,0 +1,28 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_sub_op.cc:
```diff
@@ -102,6 +102,16 @@ REGISTER_OP_CPU_KERNEL(
+    elementwise_sub_one_dnn,
+    ops::ElementwiseSubKernel<paddle::platform::CPUDeviceContext, float>,
+    ops::ElementwiseSubKernel<paddle::platform::CPUDeviceContext, double>,
+    ops::ElementwiseSubKernel<paddle::platform::CPUDeviceContext, int>,
```
As with other eltwises, do we support double, int and int64_t datatypes?

tsocha replied:
https://github.com/PaddlePaddle/Paddle/pull/36541#discussion_r797554318

jakpiase commented on paddle/fluid/framework/ir/graph_pattern_detector.cc:
```diff
@@ -918,6 +918,53 @@ PDNode *patterns::ConvActivation::operator()(
 }
 
+PDNode *patterns::ElementwiseActivation::operator()(
+    paddle::framework::ir::PDNode *elementwise_a, std::string elementwise_type,
```
I suggest using const std::string reference for these function arguments

jakpiase commented on paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.cc:
```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase commented on paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.cc:
```diff
@@ -0,0 +1,148 @@
+  std::vector<std::string> act_types = {
+      "relu", "tanh", "leaky_relu", "swish", "hardswish", "sqrt",
+      "abs",  "clip", "gelu",       "relu6", "sigmoid"};
+  std::vector<std::string> elt_types = {"elementwise_add_one_dnn",
```
Can we rename all these to use `onednn` instead of `one_dnn`?

jakpiase commented on paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.h:
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase commented on paddle/fluid/framework/ir/mkldnn/onednn_replace_pass.cc:
```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase commented on paddle/fluid/framework/ir/mkldnn/onednn_replace_pass.h:
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase commented on paddle/fluid/framework/ir/mkldnn/onednn_replace_pass.h:
```diff
@@ -0,0 +1,40 @@
+namespace ir {
+
+/*
+ * \brief   Replace PaddlePaddle operator with te custom one which supports
```
```suggestion
 * \brief   Replace PaddlePaddle operator with the custom one which supports
```

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_add_one_dnn_op.cc:
```diff
@@ -0,0 +1,87 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase replied:
```suggestion
/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_mul_one_dnn_op.cc:
```diff
@@ -0,0 +1,85 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase commented on paddle/fluid/operators/elementwise/elementwise_mul_one_dnn_op.h:
```diff
@@ -0,0 +1,29 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase commented on paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h:
```diff
@@ -32,6 +32,39 @@ using dnnl::stream;
+      }
+    }
+    return post_operations;
+  }
```
It's so nicely done! We should really unify all our `X+activation` fuse passes to use one single algo_map for that, but that's just a digression

jakpiase commented on python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_elt_act_fuse_pass.py:
```diff
@@ -0,0 +1,327 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

sfraczek commented on paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.cc:
```diff
@@ -0,0 +1,148 @@
+
+    auto *elementwise_op = elementwise->Op();
+
+    if (elementwise_op->HasAttr("use_mkldnn")) {
```
Why not check this inside pattern matching like in OneDNNOp?

tsocha replied:
It is consistent with other passes, I check it here to throw an exception with customized message.

sfraczek commented on paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.h:
```diff
@@ -0,0 +1,45 @@
+/*
+ * \brief   Fuse the Elementwise and activation operators into single
+ * OneDNN's
+ *          Elementwize with post-op.
```
Seems like an unnecessary line break

sfraczek commented on paddle/fluid/framework/ir/mkldnn/onednn_replace_pass.cc:
```diff
@@ -0,0 +1,88 @@
+    gpd(graph, handler);
+  }
+
+  // get_node_from_factory();
```
Can these comments be removed?

sfraczek commented on paddle/fluid/framework/ir/mkldnn/onednn_replace_pass.cc:
```diff
@@ -0,0 +1,88 @@
+
+    auto handler = [op_type, onednn_op_pattern](
+        const GraphPatternDetector::subgraph_t &subgraph, Graph *g) {
+      VLOG(4) << "Replace " << op_type << " with onednn version op.";
```
```suggestion
      VLOG(4) << "Replace " << op_type << " op with onednn version.";
```

sfraczek commented on paddle/fluid/operators/elementwise/elementwise_add_one_dnn_op.h:
```diff
@@ -0,0 +1,29 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```

jakpiase reviewed (dismissed):
LGTM

sfraczek reviewed (dismissed):
LGTM

lidanqing-vv commented:
@phlrain Hi, Hongyu, this is the PR that using passes to add new attributes. Please note in this PR, use_mkldnn will still be added as new attributes if it is a new mkldnn op

lidanqing-vv commented:
@tsocha Hi is this PR finalized ? And most CI are pending, is it CI's issues ?

lidanqing-vv commented:
@baoachun Could you please start your reivew ?

tsocha commented:
> @tsocha Hi is this PR finalized ? And most CI are pending, is it CI's issues ?

@lidanqing-intel
Yes, it's finished.
There were some problems with mkldnn quantizer config after operator has been renamed but it has been resolved.

wozna commented on paddle/fluid/inference/api/paddle_pass_builder.cc:
```diff
@@ -266,6 +266,8 @@ void CpuPassStrategy::EnableMKLDNN() {
              //  "fc_act_mkldnn_fuse_pass",
              "batch_norm_act_fuse_pass",              //
              "softplus_activation_mkldnn_fuse_pass",  //
+             "onednn_replace_pass",
```
These two passes should also be added to `quant2_int8_mkldnn_pass.py` here https://github.com/PaddlePaddle/Paddle/blob/9c2cee1c6c59ac440d4abd3d7ece8a5f8a140bd8/python/paddle/fluid/contrib/slim/quantization/quant2_int8_mkldnn_pass.py#L458
Because for now, this script has a separate list of passes. It is supposed to be merged in the future, but to avoid problems, we should update it.

jczaja reviewed (dismissed):
LVGTM

wozna reviewed (dismissed):
LGTM

arlesniak commented on python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_elt_act_fuse_pass.py:
```diff
@@ -0,0 +1,327 @@
+    def set_params(self):
+        self.operand = fluid.layers.elementwise_sub
+        self.act = fluid.layers.gelu
+        self.act_alpha = True
```
Should `act_alpha` be a number or boolean ?

tsocha replied:
it's a boolean for gelu because it is the first parameter(approximate) for this op: 
![image](https://user-images.githubusercontent.com/24435886/154151186-2fab248e-6f79-4c37-ae5f-53241fbc9a4a.png)
link: https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/nn/functional/gelu_en.html#gelu

arlesniak commented on python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_elt_act_fuse_pass.py:
```diff
@@ -0,0 +1,327 @@
+    def set_params(self):
+        self.operand = fluid.layers.elementwise_mul
+        self.act = fluid.layers.gelu
+        self.act_alpha = True
```
Ditto

arlesniak commented on paddle/fluid/inference/api/mkldnn_quantizer_config.cc:
```diff
@@ -54,6 +54,14 @@ MkldnnQuantizerConfig::MkldnnQuantizerConfig() {
+  rules_["elementwise_mul_onednn"]["X"] = ScaleAlgo::KL;
+  rules_["elementwise_mul_onednn"]["Y"] = ScaleAlgo::KL;
+  rules_["elementwise_mul_onednn"]["Out"] = ScaleAlgo::KL;
+
```
Don't we need `rules_` section for `elementwise_sub_onednn` as well ? What is your opinion?

tsocha replied:
There is no rule for elementwise_sub operator.
I added these rules only for ops which already have rules.

wozna replied:
I agree elementwise_sub is not quantized yet, so there is no need to add it now.

wozna reviewed (approved):
LGTM

arlesniak reviewed (dismissed):
LGTM

jakpiase reviewed (approved):
LGTM

tsocha commented:
PR-CI-BUILD passed but status isn't reported here:
![image](https://user-images.githubusercontent.com/24435886/154468442-0b8c2194-0afa-45ba-bd28-31237caf8e5e.png)

sfraczek reviewed (dismissed):
LGTM

tsocha commented:
@baoachun Could you reivew this PR?
Some other PR's depends on this code so please do it ASAP.

lidanqing-vv commented on paddle/fluid/framework/ir/mkldnn/onednn_replace_pass.cc:
```diff
@@ -0,0 +1,85 @@
+      in_out_map = {{"elementwise_add", elementwise_io_map},
+                    {"elementwise_mul", elementwise_io_map},
+                    {"elementwise_sub", elementwise_io_map}};
+
```
This map is needed, because the ops inputs and outpus are needed for find and create new onednn custom op.

lidanqing-vv commented on paddle/fluid/framework/ir/mkldnn/onednn_replace_pass.cc:
```diff
@@ -0,0 +1,85 @@
+
+      auto *replace_op = op_to_replace->Op();
+
+      replace_op->SetType(op_type + "_onednn");
```
This process will only change the type and will not change inputs and outputs

jczaja commented:
@baoachun If you have any questions or concers about this PR then please ask/let us know.  Many of tasks we already have implemented & planned are blocked by this PR . For example:
![graphviz(4)](https://user-images.githubusercontent.com/15085062/155490012-2b6fb143-6066-456b-abf2-81c373b37dc7.png)

baoachun commented:
Please add a single test based on the new single test framework.

tsocha commented:
> Please add a single test based on the new single test framework.

@baoachun 
I and @wozna have created an another solution which don't use custom operators.
https://github.com/PaddlePaddle/Paddle/pull/40081
In this ⬆️ PR we just set an attribute which is not registered anywhere and it seems to work.
I also added a test on the new test framework there.

Please tell me, should I select this alternative solution instead of custom op from this PR?

lidanqing-vv commented:
@baoachun Hi, Tomasz and Joanna made another solution that adds attributes in the passes, could you please review it #40081 Thanks!

jczaja reviewed (approved):
LGTM
